### PR TITLE
Fix multiple hooks and extensions breaking on new Spotify (Solves #845)

### PR DESCRIPTION
**Pull request by itsmeow**
This fixes multiple major hooks which were broken, the second step that #870 was missing. This PR includes #870 and solves #845.

I have tested and ran this on the newest Spotify, it works very well.

With a build of this PR Spotify should display properly again.

## PATCH DOWNLOAD + INSTRUCTIONS TO FIX FOR NEW SPOTIFY

 https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch4

## Fixed hooks by PR
 - `Spicetify.React`
 - `Spicetify.Platform`
 - `Spicetify.Player.getProgress()`
 - `Spicetify.Player.getProgressPercent()`
 - `Spicetify.ReactComponent.ContextMenu`
 - `Spicetify.ReactComponent.RightClickMenu`
 - `Spicetify.ReactComponent.Menu`

## Known currently broken hooks

 - Remove star wars easter egg
 - disable_sentry (**This is completely removed**, regardless of config settings, as of patch4. Disabling this is no longer necessary)
 - disable_logging
 - `Spicetify.Menu`
 - `Spicetify.ReactComponent.MenuItem`
 - `Spicetify.ReactComponent.AlbumMenu`
 - `Spicetify.ReactComponent.PodcastShowMenu`
 - `Spicetify.ReactComponent.ArtistMenu`
 - `Spicetify.ReactComponent.PlaylistMenu`

## Known issues

- Some apps will not work due to the above broken hooks (let me know if any hooks you use are listed above as broken!)

## loopyLoop.js Fix

patch4 currently doesn't include this, you can fix it yourself by downloading and replacing the [loopyLoop.js from the repo](https://raw.githubusercontent.com/itsmeowForks/spicetify-cli/master/Extensions/loopyLoop.js) or manually updating it as so:

Open up `spicetify-cli/Extensions/loopyLoop.js`

Change:
```js
const bar = document.querySelector(".playback-bar .progress-bar");
```

to

```js
const bar = document.querySelector(".playback-bar > div:nth-child(2)");
```

### CODE REVIEW

**Comment by itsmeow** [comment]
I will say, this isn't perfect but it's a solution for now as it appears several other hooks are still broken, including ones used by lyrics plus.
image
Here's a picture of lyrics plus not working. But the popup lyrics work! Theming seems to work (I think?)
My canvas extension has broken because the AuthorizationAPI is not on the Spicetify object anymore.

They key here is the app at least loads, so install Go, run `go build` in the root directory from my fork, and replace your spicetify with spicetify-cli.exe, and things should work! I could also post the build or make a GitHub action or something to build the PR if anyone wants.

**Comment by itsmeow** [comment]
Fixed the Spicetify.Platform hook. Some React hooks are still broken, so Lyrics Plus doesn't work. Most extensions will now work though.

**Comment by itsmeow** [comment]
Build here: https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch1 (allowed under terms of LGPL)

**Comment by blue-int** [comment]
@itsmeow Could you explain more about how to install?
I understood as:
1. download the file and move it to ~/spicetify-cli
2. rename original spicetify as spicetify-old
3. rename downloaded file as spicetify
4. Run the two lines of code

Is this right?

**Comment by itsmeow** [comment]
Yep that's exactly how it's done.

**Comment by owainjones74** [comment]
It's nice to finally see some movement with this again. Your work will not go unappreciated <3

**Comment by Delta18-Git** [comment]
This seems to work for me too! However, I am waiting for this to get merged and the Lyrics to be fixed as I need the Lyrics Plus extension, which is why I installed Spicetify in the first place.

**Comment by Delta18-Git** [comment]
This seems to work for me too! However, I am waiting for this to get merged as I need the Lyrics Plus extension, which is why I installed Spicetify in the first place.

**Comment by itsmeow** [comment]
The pop up lyrics still work!

**Comment by fugkco** [comment]
Didn't work for me. I get the following error in the dev tools:
```
xpui.js:1 Uncaught SyntaxError: Unexpected token ','
```

**Comment by itsmeow** [comment]
> 
> 
> Didn't work for me. I get the following error in the dev tools:
> 
> ```
> xpui.js:1 Uncaught SyntaxError: Unexpected token ','
> ```

What's your Spotify version?

**Comment by fugkco** [comment]
Okay, so I downgraded, then upgraded to the latest version when you sent that message to get the version number, and it sudden works. Though, I still get the actual error when applying:

```
$ spicetify backup apply enable-devtool
spicetify v
info There is available backup.
info Clear current backup:
success Backup is cleared.
Backing up app files:
OK
Extracting:
OK
Preprocessing:
OK
OK
success Everything is ready, you can start applying now!
Copying raw assets:
OK
Overwriting themed assets:
OK
Transferring user.css:
OK
Applying additional modifications:
error Cannot find symbol for Custom app React symbols
OK
success Spotify is spiced up!
success DevTool enabled!
```

**Comment by CharlieS1103** [comment]
Sorry about that question if you don't have the time but I'm just curious, what exactly was the problem with the regex and how do the hooks work?

**Comment by itsmeow** [comment]
> 
> 
> Sorry about that question if you don't have the time but I'm just curious, what exactly was the problem with the regex and how do the hooks work?

I actually appreciate this so much, I love talking about stuff like this! I find it really interesting.

The way spicetify-cli works, from what I know, is it extracts these spa packages inside Spotify's assets that basically contain the "website" source code (find in `%appdata%/Spotify/Apps/xpui`), as Spotify runs on Chromium Embedded Framework, and then they just make a website to display inside that embedded Chromium. So there's just a bunch of HTML, CSS, and JS files that get displayed. The "modification" part of spicetify involves patching the text files and replacing parts of it with different parts.

So the regex will just find a spot in the source, and insert something relevant to the function of Spicetify. For extensions, it adds script tags to the end of `index.html`, for CSS themes, it first maps all the obfuscated css class names to more readable ones, and then applies some default themes and then your theme specifically. For Javascript APIs like the React hook, Spicetify will find specific spots in `xpui.js` or `xpui~vendor.js` where relevant APIs are used and then copy them to the `Spicetify` global object.

The React hook broke because Spotify likes changing things around internally between updates, and the specific spot where the API was used no longer was formatted the same to be detected by the previous regex.

In the case of the hook you referenced, they just changed the function notation in that spot for some reason. `()=>x` and `function(){return x}` are identical in JavaScript.

The hook could then no longer find that spot in the code to insert its code, so it errored.

**Comment by CharlieS1103** [comment]
Oh, so in order to fix the remaining hooks you just need to find out what spotify changed in those specific spots? I might give it a shot, although not knowing go nor regexp might be a small disadvantage lol.

**Comment by itsmeow** [comment]
It's a bit more complicated than that, but yeah I did actually edit the replace function to print if it didn't find any matches and found quite a few broken hooks right now. The main thing that needs fixing are the react component menu item hooks and the sentry disable hook. The key is basically finding the original hook in Spotify's code, then finding a place where it's used frequently and then using regex to patch that to something that has the same function while assigning `Spicetify.SomeHook` to the value. Better the higher up in the file it is, as it will make the preprocess slower if it's far down (hence it being much slower on my patch). It's really not something I'm qualified or good enough to do, but I've managed to make the fix for the React one (it took me *hours*)

Go is pretty simple if you know JavaScript, I had no issues with picking it up. I also didn't know it.

**Comment by itsmeow** [comment]
In the case of the React hook, the original place where it was hooked quite literally didn't exist, like, it was completely removed or rewritten. So it's not like "find the old spot, it's slightly different", it's that it's completely different in some cases.

**Comment by itsmeow** [comment]
Here's a list of all broken hooks (excluding sentry + logging disable)
```
warning Hook fail: Remove star wars easter egg
warning Hook fail: Spicetify.Menu
warning Hook fail: Spicetify.ReactComponent.ContextMenu, Spicetify.ReactComponent.RightClickMenu
warning Hook fail: Spicetify.ReactComponent.Menu
warning Hook fail: Spicetify.ReactComponent.MenuItem
warning Hook fail: Spicetify.ReactComponent.AlbumMenu
warning Hook fail: Spicetify.ReactComponent.PodcastShowMenu
warning Hook fail: Spicetify.ReactComponent.ArtistMenu
warning Hook fail: Spicetify.ReactComponent.PlaylistMenu
```

**Comment by itsmeow** [comment]
It also seems `Spicetify.Player.getProgress()` and `Spicetify.Player.getProgressPercent()` are no longer functional, breaking popup lyrics and other widgets as seen by some users in #845's comments.

**Comment by itsmeow** [comment]
I've found a solution to get the progress on the new Spotify. I'll make a new release build soon.

**Comment by itsmeow** [comment]
Updated builds available: https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch2

**Comment by itsmeow** [comment]
Do note this new build requires updating the `jsHelper/spicetifyWrapper.js` file, not just the exe. I will probably have to start packaging the builds as the normal ones are, with all files included, if more files need to be changed.

**Comment by webtheo** [comment]
Can confirm that it does work on the newest patch however unfortunately lyrics-plus is still broken.
I appreciate your effort trying to fix that itsmeow, you are a legend!

**Comment by theofficialgrand1** [comment]
There is no spicetify-cli folder in ~/spicetify-cli

**Comment by theofficialgrand1** [comment]
There is only spicetify_data

**Comment by theofficialgrand1** [comment]
Never mind I figured it out, my spicetify-cli was in /usr/local/Cellar/spicetify-cli

**Comment by DavidAremaCarretero** [comment]
Thanks for your work @itsmeow ! I have seen a small bug; when you pause music, progress still counting, and then when you resume music it goes back to its playing progress.

**Comment by itsmeow** [comment]
> 
> 
> Thanks for your work @itsmeow ! I have seen a small bug; when you pause music, progress still counting, and then when you resume music it goes back to its playing progress.

Uh oh, it seems you're right. The method I wrote to get the progress is a bit janky it seems.

**Comment by itsmeow** [comment]
It seems likely another hook will have to be made to get a smooth progress, as currently the only reliable progress updates come every 100-400ms, which makes popup lyrics unable to smooth scroll and looks quite odd.

**Comment by itsmeow** [comment]
Patch 3 available fixing the issue where progress continues to increase despite music being paused
https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch3

Those already on patch 2 can just replace `jsHelper/spicetifyWrapper.js` again.
I've updated the instructions and changed the build format to be like the Releases tab on this repo, including all the files. I've also added a screenshot walkthrough for Windows and more OS info.

**Comment by CharlieS1103** [comment]
Noticed a bit of a pattern in the hooks that don't work (all of which are in preprocess.go), it seems that all, or at least a majority of them use the .createElement function. @itsmeow

**Comment by CharlieS1103** [comment]
Btw how do I edit the preprocess.go file on my computer and run it to test, I made some regex for the broken hooks but I can't really figure out how to test it @itsmeow

**Comment by afonsojramos** [comment]
@itsmeow Thank you so much for your contribution, I'll review the PR soon and provide feedback soon, thanks!
PS: Sorry for the absence, as I've been on vacation.

File: src/preprocess/preprocess.go
Comment (afonsojramos) [inline src/preprocess/preprocess.go:290]: In go, on a for range, you can just get the index by adding it to the for.
So instead of
```go
i := 0
for _, apiFunc := range splitted {
  name := re.ReplaceAllString(apiFunc, `${1}`)
  
  if strings.HasPrefix(name, "get") {
	  name = strings.Replace(name, "get", "", 1)
  }
  
  code += name + ": await " + apiFunc + ","
  code += "Spicetify.Platform[\"" + name + "\"] = v[" + fmt.Sprint(i) + "];"
  i = i + 1
}
```

```go
for apiFuncIndex, apiFunc := range splitted {
  name := re.ReplaceAllString(apiFunc, `${1}`)
  
  if strings.HasPrefix(name, "get") {
	  name = strings.Replace(name, "get", "", 1)
  }
  
  code += name + ": await " + apiFunc + ","
  code += "Spicetify.Platform[\"" + name + "\"] = v[" + fmt.Sprint(apiFuncIndex) + "];"

}
```

**Review by afonsojramos — CHANGES_REQUESTED** [review]
Other than the comment I've made, everything looks good to me!

**Comment by itsmeow** [comment]
> Btw how do I edit the preprocess.go file on my computer and run it to test, I made some regex for the broken hooks but I can't really figure out how to test it @itsmeow

You need to run `go build` and then test on spicetify-cli.exe which is output after the command.

File: src/preprocess/preprocess.go
Comment (itsmeow) [reply src/preprocess/preprocess.go:290]: Shows my inexperience with go.. Implementing!

File: src/preprocess/preprocess.go
Comment (afonsojramos) [inline src/preprocess/preprocess.go:204]: I suppose that we can leave this commented out for now, but we definitely should review it later on.

**Review by afonsojramos — APPROVED** [review]

**Comment by itsmeow** [comment]
image
I can confirm the breakage of lyrics plus is an issue with the app, not the apps *system*. A minimal test app with one react element renders fine.

**Comment by itsmeow** [comment]
Lyrics Plus is broken because `Spicetify.ReactComponent.ContextMenu` and `Spicetify.ReactComponent.Menu` are used in `OptionsMenu.js`.

I'll look into reimplementing these hooks.

**Comment by itsmeow** [comment]
I fixed it!
image

Commits + build soon!

**Comment by itsmeow** [comment]
Patch 4, fixing those hooks and thus lyrics plus:
https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch4

**Comment by saxobroko** [comment]
> Patch 4, fixing those hooks and thus lyrics plus:
> https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch4

Can confirm this solves it on Windows.

**Comment by casungo** [comment]
Can also confirm its fixed! Thanks for your work @itsmeow <3

**Comment by francescoaracu** [comment]
> Patch 4, fixing those hooks and thus lyrics plus:
> https://github.com/itsmeowForks/spicetify-cli/releases/tag/v2.5.0-patch4

Lyrics Plus is fixed but not entirely: on my installation of Spotify + Spicetify if I click on the "More" button (near Unsynced) it goes on the "Something went wrong" page. Another thing I noticed: patch4 broke New Releases custom app (it was working with patch3).

File: Extensions/loopyLoop.js
Comment (fugkco) [inline Extensions/loopyLoop.js:10]: I think `.progress-bar` was changed to `.progress-bar-wrapper` so `.playback-bar > .progress-bar-wrapper` should work and will be better if it does.

**Comment by surya27895** [comment]
still have issue with lyric plus when in at drop down mode

File: Extensions/loopyLoop.js
Comment (itsmeow) [reply Extensions/loopyLoop.js:10]: My Spotify has no such class listing on the div I was trying to target, I think the CSS map for that class is broken, I believe spicetify maps CSS to readable names in some places.

**Comment by itsmeow** [comment]
> Lyrics Plus is fixed but not entirely: on my installation of Spotify + Spicetify if I click on the "More" button (near Unsynced) it goes on the "Something went wrong" page. Another thing I noticed: patch4 broke New Releases custom app (it was working with patch3).

This can be fixed by making Spotify wider to avoid the More showing up. I think it's because the MenuItem patch is not present. Not sure how an app could've *broken* after fixing those hooks, but I suppose it's possible.

**Comment by itsmeow** [comment]
> Can also confirm its fixed! Thanks for your work @itsmeow <3
> 
> Edit: https://youtu.be/HxEa5qDTi1E sike :(
> 
> 2nd edit: After a reboot it works 🤔

I think this is an unrelated issue that was already happening where lyrics plus randomly fails to get lyrics due to "malformed JSON".

**Comment by francescoaracu** [comment]
> > Lyrics Plus is fixed but not entirely: on my installation of Spotify + Spicetify if I click on the "More" button (near Unsynced) it goes on the "Something went wrong" page. Another thing I noticed: patch4 broke New Releases custom app (it was working with patch3).
> 
> This can be fixed by making Spotify wider to avoid the More showing up. I think it's because the MenuItem patch is not present.

Correct, fixed it by disabling Netease source and reducing the right sidebar width.

> Not sure how an app could've _broken_ after fixing those hooks, but I suppose it's possible.
It's strange, maybe when you patched a hook for re-enabling Lyrics Plus it changed the behaviour of New Releases...

**Comment by Delta18-Git** [comment]
Windows Defender is not allowing me to run this fork for some reason citing Advanced ransomware protection due to NanoDefender.

**Comment by itsmeow** [comment]
> Windows Defender is not allowing me to run this fork for some reason citing Advanced ransomware protection due to NanoDefender.

Add an exception for it. It's nearly identical to a normal build, don't know why it would trigger only at the patch. You can build it yourself if you want to be safe.

**Comment by FurkanKambay** [comment]
@Delta18-Git 
>  due to NanoDefender.

just to make sure, are you sure this has something to do with this patch? sounds like Defender is screaming about the NanoDefender extension which apparently is a malware as of 10 months ago: https://www.reddit.com/r/Adblock/comments/jc447f/nano_adblocker_nano_defender_was_sold_and_should/ (check the topmost comment btw)

Stay safe.

**Comment by xXChampionsXx** [comment]
I don't know if this is because of the currently broken hooks but if I install lyrics plus my spotify just says something went wrong.
image

**Review by afonsojramos — APPROVED** [review]

**Comment by afonsojramos** [comment]
I'll merge this as it is and we'll update it if need be.

**Pull request merged**
